### PR TITLE
changes to allow injecting dns names on issuer certificate

### LINKED ISSUE #1155: 🫐 🐛 When assigning a self-signed CA cert-manager issuers to internal listeners, the operator incorrectly uses the external DNS names instead of the internal ones.
### What happened?

I assigned a self-signed CA cert-manager issuer to my internal listeners, and the operator used the wrong DNS names to request certificates from it. 
Screenshot 2024-04-04 at 6 59 51 PM

```console
redpanda@redpanda-broker-0:/$ rpk cluster info
unable to request metadata: unable to dial: tls: failed to verify certificate: x509: certificate is valid for camilo.panda.dev, *.camilo.panda.dev, not redpanda-broker-0.redpanda-broker.redpanda.svc.cluster.local.
```

### What did you expect to happen?

I expected it to use the domains under the `clusterDomain` top level domain, instead. And, `redpanda@redpanda-broker-0:/$ rpk cluster info` to work. 

### How can we reproduce it (as minimally and precisely as possible)?. Please include values file.

<details>

```console
COMPUTED VALUES:
affinity: {}
auditLogging:
  clientMaxBufferSize: 16777216
  enabled: false
  enabledEventTypes: null
  excludedPrincipals: null
  excludedTopics: null
  listener: internal
  partitions: 12
  queueDrainIntervalMs: 500
  queueMaxBufferSizePerShard: 1048576
  replicationFactor: null
auth:
  sasl:
    enabled: true
    mechanism: SCRAM-SHA-512
    secretRef: redpanda-superusers
    users: []
clusterDomain: cluster.local
commonLabels:
  azure.workload.identity/use: "true"
config:
  cluster:
    default_topic_replications: 3
    minimum_topic_replications: 3
  node:
    crash_loop_limit: 5
  pandaproxy_client: {}
  rpk: {}
  schema_registry_client: {}
  tunable:
    compacted_log_segment_size: 67108864
    group_topic_partitions: 16
    kafka_batch_max_bytes: 1048576
    kafka_connection_rate_limit: 1000
    log_segment_size: 134217728
    log_segment_size_max: 268435456
    log_segment_size_min: 16777216
    max_compacted_log_segment_size: 536870912
    topic_partitions_per_shard: 1000
connectors:
  deployment:
    create: false
  enabled: false
  test:
    create: false
console:
  config: {}
  configmap:
    create: false
  deployment:
    create: false
  enabled: false
  secret:
    create: false
enterprise:
  license: ""
  licenseSecretRef:
    key: license
    name: redpanda-license
external:
  addresses:
  - $PREFIX_TEMPLATE
  domain: camilo.panda.dev
  enabled: true
  externalDns:
    enabled: true
  prefixTemplate: rp${POD_ORDINAL}-$(echo -n $HOST_IP_ADDRESS | sha256sum | head -c
    7)
  service:
    enabled: true
  type: NodePort
fullnameOverride: ""
image:
  pullPolicy: IfNotPresent
  repository: docker.redpanda.com/redpandadata/redpanda
  tag: v23.3.7
imagePullSecrets: []
license_key: ""
license_secret_ref: {}
listeners:
  admin:
    external:
      admin-api:
        advertisedPorts:
        - 30644
        authenticationMethod: sasl
        enabled: false
        port: 30644
        tls:
          cert: letsencrypt
          enabled: true
          requireClientAuth: false
      default:
        advertisedPorts:
        - 31644
        port: 9645
        tls:
          cert: external
    port: 9644
    tls:
      cert: selfsigned
      enabled: false
      requireClientAuth: false
  http:
    authenticationMethod: http_basic
    enabled: true
    external:
      default:
        advertisedPorts:
        - 30082
        authenticationMethod: null
        port: 8083
        tls:
          cert: external
          requireClientAuth: false
      http-proxy:
        advertisedPorts:
        - 31082
        authenticationMethod: http_basic
        enabled: true
        port: 31082
        tls:
          cert: letsencrypt
          enabled: true
          requireClientAuth: false
    kafkaEndpoint: default
    port: 8082
    prefixTemplate: http-proxy$POD_ORDINAL
    tls:
      cert: selfsigned
      enabled: true
      requireClientAuth: false
  kafka:
    authenticationMethod: sasl
    external:
      default:
        advertisedPorts:
        - 31092
        authenticationMethod: null
        port: 9094
        tls:
          cert: external
      kafka-api:
        advertisedPorts:
        - 32092
        authenticationMethod: sasl
        enabled: true
        port: 32092
        tls:
          cert: letsencrypt
          requireClientAuth: false
    port: 9092
    prefixTemplate: kafka-api$POD_ORDINAL
    tls:
      cert: selfsigned
      requireClientAuth: false
  rpc:
    port: 33145
    tls:
      cert: selfsigned
      requireClientAuth: false
  schemaRegistry:
    authenticationMethod: http_basic
    enabled: true
    external:
      default:
        advertisedPorts:
        - 30081
        authenticationMethod: null
        port: 8084
        tls:
          cert: external
          requireClientAuth: false
      schema-registry:
        advertisedPorts:
        - 31081
        authenticationMethod: http_basic
        enabled: true
        port: 31081
        tls:
          cert: letsencrypt
          requireClientAuth: false
    kafkaEndpoint: default
    port: 8081
    tls:
      cert: selfsigned
      requireClientAuth: false
logging:
  logLevel: debug
  usageStats:
    clusterId: 9m4e2mr0ui3e8a215n4g
    enabled: true
monitoring:
  enabled: false
  labels: {}
  scrapeInterval: 30s
  tlsConfig: {}
nameOverride: ""
nodeSelector: {}
post_install_job:
  affinity: {}
  enabled: true
post_upgrade_job:
  affinity: {}
  enabled: true
rackAwareness:
  enabled: true
  nodeAnnotation: topology.kubernetes.io/zone
rbac:
  annotations: {}
  enabled: false
resources:
  cpu:
    cores: "15"
  memory:
    container:
      max: 105Gi
      min: 105Gi
    enable_memory_locking: true
serviceAccount:
  annotations:
    azure.workload.identity/client-id: 7bef7aa7-cb90-4860-834f-6d92e070cffc
  create: true
  name: id-rpcloud-9m4e2mr0ui3e8a215n4
statefulset:
  additionalRedpandaCmdFlags:
  - --abort-on-seastar-bad-alloc
  - --dump-memory-diagnostics-on-alloc-failure-kind=all
  annotations: {}
  budget:
    maxUnavailable: 1
  extraVolumeMounts: ""
  extraVolumes: ""
  initContainerImage:
    repository: busybox
    tag: latest
  initContainers:
    configurator:
      extraVolumeMounts: ""
      resources: {}
    extraInitContainers: ""
    fsValidator:
      enabled: true
      expectedFS: xfs
      extraVolumeMounts: ""
      resources: {}
    setDataDirOwnership:
      enabled: true
      extraVolumeMounts: ""
      resources: {}
    setTieredStorageCacheDirOwnership:
      extraVolumeMounts: ""
      resources: {}
    tuning:
      extraVolumeMounts: ""
      resources: {}
  livenessProbe:
    failureThreshold: 3
    initialDelaySeconds: 10
    periodSeconds: 10
  nodeSelector:
    cloud.redpanda.com/role: redpanda
  podAffinity: {}
  podAntiAffinity:
    custom: {}
    topologyKey: kubernetes.io/hostname
    type: hard
    weight: 100
  priorityClassName: ""
  readinessProbe:
    failureThreshold: 3
    initialDelaySeconds: 1
    periodSeconds: 10
    successThreshold: 1
  replicas: 3
  securityContext:
    allowPrivilegeEscalation: false
    fsGroup: 101
    fsGroupChangePolicy: OnRootMismatch
    runAsGroup: 65530
    runAsNonRoot: true
    runAsUser: 65530
  sideCars:
    configWatcher:
      enabled: true
      extraVolumeMounts: ""
      resources: {}
      securityContext: {}
    controllers:
      createRBAC: true
      enabled: false
      healthProbeAddress: :8085
      image:
        repository: docker.redpanda.com/redpandadata/redpanda-operator
        tag: v2.1.10-23.2.18
      metricsAddress: :9082
      resources: {}
      run:
      - all
      securityContext: {}
  startupProbe:
    failureThreshold: 120
    initialDelaySeconds: 1
    periodSeconds: 10
  terminationGracePeriodSeconds: 90
  tolerations:
  - effect: NoSchedule
    key: cloud.redpanda.com/role
    operator: Equal
    value: redpanda
  topologySpreadConstraints:
  - maxSkew: 1
    topologyKey: topology.kubernetes.io/zone
    whenUnsatisfiable: ScheduleAnyway
  updateStrategy:
    type: RollingUpdate
storage:
  hostPath: ""
  persistentVolume:
    annotations: {}
    enabled: true
    labels: {}
    nameOverwrite: ""
    size: 4096Gi
    storageClass: local-path
  tiered:
    config:
      cloud_storage_access_key: ""
      cloud_storage_api_endpoint: ""
      cloud_storage_azure_container: null
      cloud_storage_azure_shared_key: null
      cloud_storage_azure_storage_account: null
      cloud_storage_bucket: ""
      cloud_storage_cache_size: 5368709120
      cloud_storage_credentials_source: config_file
      cloud_storage_enable_remote_read: true
      cloud_storage_enable_remote_write: true
      cloud_storage_enabled: false
      cloud_storage_region: ""
      cloud_storage_secret_key: ""
    credentialsSecretRef:
      accessKey:
        configurationKey: cloud_storage_access_key
      secretKey:
        configurationKey: cloud_storage_secret_key
    hostPath: ""
    mountType: persistentVolume
    persistentVolume:
      annotations: {}
      labels: {}
      storageClass: local-path
tests:
  enabled: true
tls:
  certs:
    default:
      caEnabled: true
    external:
      caEnabled: true
    letsencrypt:
      caEnabled: false
      duration: 43800h0m0s
      issuerRef:
        kind: ClusterIssuer
        name: letsencrypt-dns
    selfsigned:
      caEnabled: true
      duration: 43800h0m0s
      issuerRef:
        kind: ClusterIssuer
        name: redpanda.local
  enabled: true
tolerations: []
tuning:
  tune_aio_events: false

```

</details>

### Anything else we need to know?

_No response_

### Which are the affected charts?

Redpanda, Operator

### Chart Version(s)

<details>

```console
❯ helm -n redpanda list
NAME             	NAMESPACE	REVISION	UPDATED                                	STATUS  	CHART          	APP VERSION
redpanda         	redpanda 	2       	2024-04-04 18:10:39.617641 -0400 EDT   	deployed	redpanda-0.1.1 	0.1.0
redpanda-broker  	redpanda 	2       	2024-04-04 22:10:41.704565747 +0000 UTC	deployed	redpanda-5.7.37	v23.3.10
redpanda-operator	redpanda 	1       	2024-04-01 18:14:29.732128 -0400 EDT   	deployed	operator-0.4.20	v2.1.15-23.3.7
```

</details>

### Cloud provider

<details>
Azure / AKS
</details>

### PR DESCRIPTION
**Pull request by alejandroEsc**
Fixes #1155

### CODE REVIEW

File: charts/redpanda/templates/certs.yaml
Comment (c4milo) [inline charts/redpanda/templates/certs.yaml:40]: true?

Comment (alejandroEsc) [reply charts/redpanda/templates/certs.yaml:40]: i am sorry, what do you mean? is the logic not what you wanted?

**Comment by alejandroEsc** [comment]
Before:
```
---
# Source: redpanda/templates/certs.yaml
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: redpanda-selfsigned-cert
  namespace: "redpanda"
  labels:
    app.kubernetes.io/component: redpanda
    app.kubernetes.io/instance: redpanda
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: redpanda
    helm.sh/chart: redpanda-5.7.37
spec:
  dnsNames:
  duration: 43800h0m0s
  isCA: false
  secretName: redpanda-selfsigned-cert
  privateKey:
    algorithm: ECDSA
    size: 256
  issuerRef:
    kind: ClusterIssuer
    name: redpanda.local
    group: cert-manager.io
```
and
```
---
# Source: redpanda/templates/certs.yaml
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: redpanda-letsencrypt-cert
  namespace: "redpanda"
  labels:
    app.kubernetes.io/component: redpanda
    app.kubernetes.io/instance: redpanda
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: redpanda
    helm.sh/chart: redpanda-5.7.37
spec:
  dnsNames:
  duration: 43800h0m0s
  isCA: false
  secretName: redpanda-letsencrypt-cert
  privateKey:
    algorithm: ECDSA
    size: 256
  issuerRef:
    kind: ClusterIssuer
    name: letsencrypt-dns
    group: cert-manager.io
```

And now:
```
---
# Source: redpanda/templates/certs.yaml
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: redpanda-selfsigned-cert
  namespace: "redpanda"
  labels:
    app.kubernetes.io/component: redpanda
    app.kubernetes.io/instance: redpanda
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: redpanda
    helm.sh/chart: redpanda-5.7.37
spec:
  dnsNames:
    - redpanda-cluster.redpanda.redpanda.svc.cluster.local
    - redpanda-cluster.redpanda.redpanda.svc
    - redpanda-cluster.redpanda.redpanda
    - "*.redpanda-cluster.redpanda.redpanda.svc.cluster.local"
    - "*.redpanda-cluster.redpanda.redpanda.svc"
    - "*.redpanda-cluster.redpanda.redpanda"
    - redpanda.redpanda.svc.cluster.local
    - redpanda.redpanda.svc
    - redpanda.redpanda
    - "*.redpanda.redpanda.svc.cluster.local"
    - "*.redpanda.redpanda.svc"
    - "*.redpanda.redpanda"
  duration: 43800h0m0s
  isCA: false
  secretName: redpanda-selfsigned-cert
  privateKey:
    algorithm: ECDSA
    size: 256
  issuerRef:
    kind: ClusterIssuer
    name: redpanda.local
    group: cert-manager.io
```
and
```
---
# Source: redpanda/templates/certs.yaml
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: redpanda-letsencrypt-cert
  namespace: "redpanda"
  labels:
    app.kubernetes.io/component: redpanda
    app.kubernetes.io/instance: redpanda
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: redpanda
    helm.sh/chart: redpanda-5.7.37
spec:
  dnsNames:
    - redpanda-cluster.redpanda.redpanda.svc.cluster.local
    - redpanda-cluster.redpanda.redpanda.svc
    - redpanda-cluster.redpanda.redpanda
    - "*.redpanda-cluster.redpanda.redpanda.svc.cluster.local"
    - "*.redpanda-cluster.redpanda.redpanda.svc"
    - "*.redpanda-cluster.redpanda.redpanda"
    - redpanda.redpanda.svc.cluster.local
    - redpanda.redpanda.svc
    - redpanda.redpanda
    - "*.redpanda.redpanda.svc.cluster.local"
    - "*.redpanda.redpanda.svc"
    - "*.redpanda.redpanda"
  duration: 43800h0m0s
  isCA: false
  secretName: redpanda-letsencrypt-cert
  privateKey:
    algorithm: ECDSA
    size: 256
  issuerRef:
    kind: ClusterIssuer
    name: letsencrypt-dns
    group: cert-manager.io
```

File: charts/redpanda/templates/certs.yaml
Comment (alejandroEsc) [reply charts/redpanda/templates/certs.yaml:40]: OH, ok , yes we do not want that to be default for the or statement, as it would place those entries and override current behavior thus changing expectations (breaking change). We would like to keep things as we have them and provide dials for changing behavior, like in this case.

File: charts/redpanda/values.yaml
Comment (chrisseto) [inline charts/redpanda/values.yaml:223]: Any reason to not include the actual value here instead of commenting it out?

We'll want the type of the values struct to match the type of `values.yaml`. If we don't include the value in `values.yaml`, then `ApplyInternalDNSNames` should be a pointer to a bool (`*bool`) to indicate that `nil` is an acceptable value.

File: charts/redpanda/values.go
Comment (chrisseto) [inline charts/redpanda/values.go:416]: nit: How about `includeInternalDNSNames`? I don't think `Apply` has a very understandable meaning in this context, do you?

File: charts/redpanda/values.yaml
Comment (chrisseto) [inline charts/redpanda/values.yaml:222]:
```suggestion
      # if you wish to have Kubernetes internal dns names (IE the headless service of the redpanda StatefulSet) included in `dnsNames` of the  certificate even, when supplying an issuer.
```

File: charts/redpanda/Chart.yaml
Comment (chrisseto) [inline charts/redpanda/Chart.yaml:26]: A linter will complain that you haven't updated the README.md. I don't think I've made a task for that yet but you can find the command to run under the `ci:lint` task definition.

File: charts/redpanda/values.go
Comment (alejandroEsc) [reply charts/redpanda/values.go:416]: Include would mean add in addition to. Apply means to place. I think apply makes more sense than include since no additional dnsNames are added.

File: charts/redpanda/values.yaml
Comment (alejandroEsc) [reply charts/redpanda/values.yaml:223]: yes its a pointer, there is no need for this to be included in the api, it did not exist before and we are using dig so we can ignore it if missing.

File: charts/redpanda/values.go
Comment (chrisseto) [reply charts/redpanda/values.go:416]: Aren't the external DNS names "added" to the list? lines 54-57 in certs.yaml (Can't link to non-diffs in a PR :( )

We could also consult our user facing language expert @JakeSCahill ?

File: charts/redpanda/values.yaml
Comment (JakeSCahill) [inline charts/redpanda/values.yaml:241]:
```suggestion
      # Enabling `applyInternalDNSNames` for external certificates ensures that internal services
      # can still validate the internal DNS names (from the headless Service of the Redpanda StatefulSet).
      # applyInternalDNSNames: false
```

File: charts/redpanda/values.yaml
Comment (JakeSCahill) [inline charts/redpanda/values.yaml:222]:
```suggestion
      # Ensures Kubernetes internal DNS names (from the headless Service of the Redpanda StatefulSet) are included in the certificate's SAN (Subject Alternative Names) list.
      # This option is useful when encountering TLS verification errors due to mismatched DNS names.
      # Example error without this option: "tls: failed to verify certificate: x509: certificate is valid for custom.domain, not redpanda-broker-0.redpanda.svc.cluster.local".
```

**Review by JakeSCahill — COMMENTED** [review]
I just reviewed the user-facing side of things in the `values.yaml`

Looks like a good change. I had one question though.

File: charts/redpanda/values.yaml
Comment (alejandroEsc) [reply charts/redpanda/values.yaml:223]: Yes, but not for external issuers, which you may want to inject these names. We do not do that by default today which is why i set it to false.

Comment (JakeSCahill) [reply charts/redpanda/values.yaml:223]: OK. Is this just for Issuers or also for Secrets?

File: charts/redpanda/values.yaml
Comment (chrisseto) [reply charts/redpanda/values.yaml:225]: I don't think I understand your reasoning for not including this with a default value of `false`.

As the field exists on our values struct and therefore JSONSchema, it's part of the API. Including the default in values.yaml is a backward compatible change as the users' values will be merged with it. Including the default also makes it much easier to reason about the the behavior because there are two states (true / false) vs three states (true / false / nil).

**Review by chrisseto — APPROVED** [review]
LGTM once CI is passing.

Please squash your commits before merging 😄 (Either rebase locally or use github's squash & merge).

**Comment by RafalKorepta** [comment]
CI will not pass as @alejandroEsc have `some-users-updated.yaml` that comes from special workflow that tests SASL users upgrades. It would be good to move that to go base test suite.

**Pull request merged**
